### ISSUE TITLE
Store original filename in metadata associated with layer instances

### ISSUE THREAD

**Issue by haesleinhuepf** [napari/napari#2033]
## 🚀 Feature
It would be cool if layers containing image data which were directly loaded from disc have an attribute/property containing meta data, especially the filename. 

## Motivation
I'm working on a [code generator](https://github.com/clEsperanto/napari_pyclesperanto_assistant/blob/master/napari_pyclesperanto_assistant/_scriptgenerators/_ScriptGeneratorBase.py#L4) which takes a list of napari-layers and generates code for reproducing their content. Starting point in my plugin is a layer that has the attribute "filename".

## Pitch
By storing meta data, such as the original filename, in layers we can use it to document what has been done with images within napari. 

## Alternatives
~Attaching a general-purpose metadata dictionary to layers which can be manipulated by any plugin would be most flexible.~
Edit: @kevinyamauchi pointed me to the existing metadata dictionary. Thanks!

## Additional context
I think one could build this in [here](https://github.com/napari/napari/blob/e61e4fc85b94597b32952599e565330e66d42f1f/napari/components/viewer_model.py#L969). If we conclude on how to do it, I'd be happy to provide PR.

**Comment by haesleinhuepf** [comment]
I dedicate this issue to an absorbing [computer game](https://en.wikipedia.org/wiki/Metro_2033_(video_game))

**Comment by kevinyamauchi** [comment]
Hey @haesleinhuepf ! We have a `metadata` attribute on all layers (see [here ](https://github.com/napari/napari/blob/e61e4fc85b94597b32952599e565330e66d42f1f/napari/layers/base/base.py#L32-L33)on the base layer). It stores a dictionary that is settable by the user. As far as I know, It doesn't have any "official" uses now and is there for the user to use as they see fit.  I could imagine this being useful for filenames as well as other provenance/logging. Would this help your use case?

```python
# create a layer with metadata
im_layer = viewer.add_image(im_data, metadata={'filename': 'amazing_image.tif'}

# get the metadata
metadata = im_layer.metadata

# update and save the metadata
new_metadata = {'processing_step_1': {'filter': 'gaussian_low_pass', 'sigma': [1, 1, 1]}}
metadata.update(new_metadata)
im_layer.metadata = new_metadata

```

**Comment by haesleinhuepf** [comment]
> Would this help your use case?

Yes, perfect! Thanks :-) I will use this. 

How about writing in the filename after loading? Does anything speak against doing this per default?

**Comment by haesleinhuepf** [comment]
Furthermore, would it make sense to have `metadata` per default as empty dictionary `{}` instead of `None`? That would make it more accessible.

https://github.com/napari/napari/blob/e61e4fc85b94597b32952599e565330e66d42f1f/napari/components/viewer_model.py#L655

**Comment by kevinyamauchi** [comment]
> Yes, perfect! Thanks :-) I will use this.

Awesome!

> How about writing in the filename after loading? Does anything speak against doing this per default?

I am open to it. If we make it a default, I think it should be always overridden by user and plugin inputs. I could imagine that different people have different ideas about how much of the path they'd like to save (e.g., base name vs. full path) and then there's the issue of `\` vs `/`. I am curious about what others think!
> Furthermore, would it make sense to have metadata per default as empty dictionary {} instead of None? That would make it more accessible.

Ah sorry, that's a little confusing. I am pretty sure that in the end, it does default to an empty dictionary (i.e., `{}`) since this is how it gets saved in the layer base class (i.e., in the default case `metadata` is passed as `None` to the layer `__init__` method and then the `__init__` method saves it as `{}`):

https://github.com/napari/napari/blob/e61e4fc85b94597b32952599e565330e66d42f1f/napari/layers/base/base.py#L177

**Comment by haesleinhuepf** [comment]
> I am pretty sure that in the end, it does default to an empty dictionary (i.e., `{}`) since this is how it gets saved in the layer base class

Correct! I need to read more napari code ;-) Thanks for your support. My code looks much cleaner now :-)

**Comment by kevinyamauchi** [comment]
Awesome! I am looking forward to checking out your code generator.

**Comment by sofroniewn** [comment]
We have the [viewer.open](https://github.com/napari/napari/blob/v0.4.2/napari/components/viewer_model.py#L781) method that take a path to add a layer and we could automatically store that path in Layer `metadata`. Would that help @haesleinhuepf?

**Comment by haesleinhuepf** [comment]
> Would that help @haesleinhuepf?

How about implementing it [here](https://github.com/napari/napari/blob/e61e4fc85b94597b32952599e565330e66d42f1f/napari/components/viewer_model.py#L969)? It would then work with all loader plugins, I think.

**Comment by jni** [comment]
I like this proposal and where it's going, but it's important to note that some plugins may want to load multiple files into a single layer. For example, a list of tiffs might become a single image layer, or a list of csvs might become a single tracks layer. See #1883 for details. So there's a bit of complexity to this.

But, as long as we take this into account, I'm very much in favour of making this kind of metadata built-in.

**Comment by sofroniewn** [comment]
Yeah, I agree with @jni we should consider some of those corner cases, but otherwise this then seems very reasonable

**Comment by haesleinhuepf** [comment]
If we call the meta data entry `source` or similar, it can be a filename, a list of files or a directory...

**Comment by jni** [comment]
Or a URL! I very much like `source` as a suggestion! 😊

**Comment by haesleinhuepf** [comment]
Or a `function` such as `skimage.data.camera`! 🙂

**Comment by lucyleeow** [comment]
Can this be closed by #2518 ?

**Comment by DragaDoncila** [comment]
@lucyleeow it sure can!
